almighty/almighty-core
# Update user profile info in Keycloak

--- Commit message ----

1. Updates platform SSO Keycloak with user information on getting a PATCH User request.
2. Updates platform SSO keycloak with user attributes without deleting the ones which platform backend is not aware of.

Fixes #877 

---End-----

API is now working as expected, only tests need to be added.

- [x] Tests for keycloak interaction
- [x] Move out hardcoded url into configuration.
- [x] API call for Update User needs to be update KC as well.
- [x] Ensure 'extra' keycloak account attributes are not deleted.


Steps to test:

1. Run core build `./bin/alm`
2. Navigate to `localhost:8080/api/login/authorize`
3. Complete auth flow and get the token ( lets call it "ALM_TOKEN" ) from the URL.
4. Call update User API:
```
./bin/alm-cli update users --key $ALM_TOKEN  --payload '{ "data":{ "type":"idenities" ,  "attributes":{"fullName":"hi","contextInformation":{ "recentContexts": [{"user":"rhn-support-pmuir","space":null},{"user":"rhn-support-pmuir","space":"pete123456"}],"recentSpaces": ["5004693d-eaf0-4929-a7e9-e92caf67391e","5004693d-eaf0-4929-a7e9-e92caf67391f"]  },"bio":"something222"}}}'  -H localhost:8080 --pp


2017/03/22 17:16:59 [INFO] started id=HoU4Xhhu PATCH=http://localhost:8080/api/users
2017/03/22 17:17:00 [INFO] completed id=HoU4Xhhu status=200 time=1.312241139s


{
    "data": {
        "attributes": {
            "bio": "something222",
            "contextInformation": {
                "recentContexts": [
                    {
                        "space": null,
                        "user": "rhn-support-pmuir"
                    },
                    {
                        "space": "pete123456",
                        "user": "rhn-support-pmuir"
                    }
                ],
                "recentSpaces": [
                    "5004693d-eaf0-4929-a7e9-e92caf67391e",
                    "5004693d-eaf0-4929-a7e9-e92caf67391f"
                ]
            },
            "email": "d@d.com",
            "fullName": "hi",
            "imageURL": "https://www.gravatar.com/avatar/a937f72ddab021504b148a7d984bf4cb.jpg",
            "providerType": "kc",
            "url": "",
            "username": "sbose78@gmail.com"
        },
        "id": "8fa0e8a4-4dd3-4c27-901a-4fc98aca36a4",
        "links": {
            "self": "http://localhost:8080/api/users/8fa0e8a4-4dd3-4c27-901a-4fc98aca36a4"
        },
        "type": "identities"
    }
}
```

5. Verify on Keycloak side.

```
GET http://sso.prod-preview.openshift.io/auth/realms/fabric8-test/account
Authorization: Bearer <TOKEN>
Accept: application/json, text/plain, */*
Content-Type: application/json

{
"id": "8fa0e8a4-4dd3-4c27-901a-4fc98aca36a4",
"createdTimestamp": 1490095538157,
"username": "sbose78@gmail.com",
"enabled": true,
"totp": false,
"emailVerified": false,
"firstName": "hi",
"lastName": "",
"attributes": {
"bio": [
  "something222"
],
},
"disableableCredentialTypes": [],
"requiredActions": [],
}
```
